src/runtime/current_thread/runtime.rs

```diff
@@ -0,0 +1,111 @@
+use executor::current_thread::{self, CurrentThread};
+
+use tokio_reactor::{self, Reactor};
+use tokio_timer::timer::{self, Timer};
+use tokio_executor;
+
+use futures::Future;
+
+use std::io;
+
+/// Single-threaded runtime provides a way to start reactor
+/// and executor on the same thread.
+///
+/// The single-threaded runtime is not `Send` and cannot be
+/// safely moved to other threads.
+///
+/// # Examples
+///
+/// Creating a new `Runtime` with default configuration values.
+///
+/// ```
+/// use tokio::runtime::current_thread::Runtime;
+/// use tokio::prelude::*;
+///
+/// let mut runtime = Runtime::new().unwrap();
+///
+/// // Use the runtime...
+/// // runtime.block_on(f); // where f is a future
+/// ```
+#[derive(Debug)]
+pub struct Runtime {
+    reactor_handle: tokio_reactor::Handle,
+    timer_handle: timer::Handle,
+    executor: CurrentThread<Timer<Reactor>>,
+}
+
+impl Runtime {
+    /// Returns a new single-threaded runtime initialized with default
+    /// configuration values.
+    pub fn new() -> io::Result<Runtime> {
+        // We need a reactor to receive events about IO objects from kernel
+        let reactor = Reactor::new()?;
+        let reactor_handle = reactor.handle();
+
+        // Place a timer wheel on top of the reactor. If there are no timeouts to fire, it'll let the
+        // reactor pick up new some new external events.
+        let timer = Timer::new(reactor);
+        let timer_handle = timer.handle();
+
+        // And now put a single-threaded executor on top of the timer. When there are no futures ready
+        // to do something, it'll let the timer or the reactor to generate some new stimuli for the
+        // futures to continue in their life.
+        let executor = CurrentThread::new_with_park(timer);
+
+        let runtime = Runtime { reactor_handle, timer_handle, executor };
+        Ok(runtime)
+    }
+
+    /// Blocks and runs the given future.
+    ///
+    /// This is similar to running a runtime, but uses only the current thread.
+    pub fn block_on<F: Future<Item = (), Error = ()>>(&mut self, f: F)
```

Comment:
I do not believe that `Item` or `Error` need to be specified here. Any types can work. So, this should be switched to `F: Future`.

Reply:
I did it because tokio::run accepts the same type. So any time you call tokio::run you may call `block_on` without changing the code

Reply:
But the point of `block_on` is that you can get the result of the passed future back. As it blocks until that future is resolved

The normal runtime does not have such a thing

Reply:
Also, the concurrent runtime will get block_on, it is just tricker to implement. I believe that @danburkert is working on this.

Reply:
See #328.  I think the API should match here, except it won't need the `Send` bound, iirc.

Reply:
@danburkert right. done